src/integration.ts

```diff
@@ -0,0 +1,146 @@
+import { CallExpr } from "./expression";
+import { ASL, State } from "./asl";
+import { FunctionlessNode } from "./node";
+import { VTL } from "./vtl";
+import { AnyFunction } from "./util";
+
+/**
+ * All integration methods supported by functionless.
+ */
+interface IntegrationMethods {
+  /**
+   * Integrate with VTL applications like AppAsync.
+   * @private
+   */
+  vtl: (call: CallExpr, context: VTL) => string;
+  /**
+   * Integrate with ASL applications like StepFunctions.
+   * @private
+   */
+  asl: (call: CallExpr, context: ASL) => Omit<State, "Next">;
+}
+
+/**
+ * Integration types supported by Functionless.
+ *
+ * Add an integration by creating any object that has a property named "kind" and either implements the
+ * {@link IIntegration} interface or has methods that implement it (or both).
+ *
+ * Example showing both strategies:
+ *
+ * ```ts
+ * export class Function implements IntegrationHandler {
+ *    readonly kind = "Function",
+ *
+ *    // Integration Handler for ASL
+ *    public asl(call, context) {
+ *       // return Step Function task.
+ *    }
+ *
+ *    // Example class method - some wrapper function that generates special ASL tasks when using a Function.
+ *    public specialPayload = makeIntegration<() => string>({
+ *        kind: "Function.default",
+ *        asl: (call, context) => {
+ *            // return step function task
+ *        }
+ *    });
+ * }
+ *
+ * // an interface to provide the actual callable methods to users
+ * export interface Function {
+ *    // call me to send a string payload
+ *    (payload: String) => string
+ * }
+ *
+ * // use
+ *
+ * const func1 = new Function(...);
+ * // uses the ASL
+ * new StepFunction(..., () => {
+ *    func1("some string");
+ *    // Calling our special method in a step function closure
+ *    func1.specialPayload();
+ * })
+ * ```
+ *
+ * If an integration does not support an integration type, leave the function undefined or throw an error.
+ *
+ * Implement the unhandledContext function to customize the error message for unsupported contexts.
+ * Otherwise the error will be: `${this.name} is not supported by context ${context.kind}.`
+ */
+export type IIntegration = {
```

Comment:
On second thought, should we just call this `Integration` for simplcity? The class, `Integration`, can be called `IntegrationImpl` since it'll never be seen by consumers.